[IMPROVEMENT] Pipefy - Create Card requires user to input GraphQL query

michelle0927 opened the pull request:
Resolves #4194 

Updates the Pipefy Create Card action to list the available fields as additionalProps instead of requiring the user to enter a GraphQL query.

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4311"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

dylburger commented:
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

andrewjschuang commented on components/pipefy/pipefy.app.mjs:
```diff
@@ -0,0 +1,236 @@
+        url: `${this._getBaseUrl()}/${endpoint}`,
+        headers: this._getHeaders(),
+        data,
+      }; console.log(config);
```
```suggestion
      };
```

andrewjschuang commented on components/pipefy/pipefy.app.mjs:
```diff
@@ -0,0 +1,236 @@
+    async _makeQueriesRequest(data = null) {
+      return this._makeRequest(data, "queries");
+    },
+    async _makeGraphQlRequest(data = null) {
```
No need for this initialization

```suggestion
    async _makeGraphQlRequest(data) {
```

andrewjschuang commented on components/pipefy/pipefy.app.mjs:
```diff
@@ -0,0 +1,236 @@
+      }; console.log(config);
+      return axios($, config);
+    },
+    async _makeQueriesRequest(data = null) {
```
No need for this initialization

```suggestion
    async _makeQueriesRequest(data) {
```

andrewjschuang commented on components/pipefy/actions/create-card/create-card.mjs:
```diff
@@ -35,17 +74,34 @@ export default {
-      },
-      data: this.graphql_query,
-    });
+    return await this.pipefy._makeRequest(data, "graphql");
```
```suggestion
    return await this.pipefy._makeRequest(data, "graphql", $);
```

Actually, creating a method and calling it here would be better

andrewjschuang commented on components/pipefy/actions/create-card/create-card.mjs:
```diff
@@ -35,17 +74,34 @@ export default {
 
-    if (!this.graphql_query) {
-      throw new Error("Must provide graphql_query parameter.");
+    let mutationString = `
```
Is there a better way to create a query with variables?
Maybe like in **Shopify Partner** or **Shopify**?

andrewjschuang reviewed (changes requested):
Hi @michelle0927, looking very good, I've added just a few comments

andrewjschuang reviewed (dismissed):
Thanks, looks great to me!

vellames reviewed (dismissed):
lgtm

vunguyenhung commented:
Hello @michelle0927, I have tested this PR and there's only 1 test cases failed. Then it's ready for release.

Please check the test report below for more information
[Pipefy_4311_2202.pdf](https://github.com/PipedreamHQ/pipedream/files/9675861/Pipefy_4311_2202.pdf)

Great job!

michelle0927 commented:
@vunguyenhung I created a Pipe using the Recruitment Process template, and confirmed that the "recruiter" field is expecting an `id` instead of a `string`. If you enter the `id` of a member/user, you shouldn't see that error. I added the field `type` to the prop descriptions to give the user a better idea of what to enter.

I also discovered that the pipe's "start_fields" are required to create a card, so I added those fields in additionalProps as well. (Hint: Candidate expects an id as well).

vunguyenhung commented:
Hello @michelle0927, thank you for explaining! I'm have test the Create Card action again following your explanation and it passed! Ready for release!

New Test Report
[Pipefy_4311_2031.pdf](https://github.com/PipedreamHQ/pipedream/files/9684572/Pipefy_4311_2031.pdf)

michelle0927 commented:
/approve
